dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+// This is free and unencumbered software released into the public domain.
+//
+// Anyone is free to copy, modify, publish, use, compile, sell, or
+// distribute this software, either in source code form or as a compiled
+// binary, for any purpose, commercial or non-commercial, and by any
+// means.
+//
+// In jurisdictions that recognize copyright laws, the author or authors
+// of this software dedicate any and all copyright interest in the
+// software to the public domain. We make this dedication for the benefit
+// of the public at large and to the detriment of our heirs and
+// successors. We intend this dedication to be an overt act of
+// relinquishment in perpetuity of all present and future rights to this
+// software under copyright law.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+#pragma once
+#ifndef FLAT_GLTF_H
+#define FLAT_GLTF_H
+
+#ifndef FLATBUFFERS_IDL_H_
+#error Include flatbuffers/idl.h before flat_gltf.h so glTF models can be loaded.
+#endif
+
+#include <functional>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include "glincludes.h"
```

Comment:
Any reason this is header only? I'd prefer this as `glloadgltf.cpp` so it fits in with `glloadiqm.cpp` and the rest of the gl*.cpp` files. That would also simplify the question of where headers go :)

Reply:
I originally named `gltfwrapper.cpp` `glloadgltf.cpp`, but it also contains rendering code, so it does more than just loading. So there is the file naming question. There is also the question of how much code is reasonable to go in a single file. STB libraries go up to about 3000 lines, which is still maintainable, but source files in the Lobster codebase seem to be somewhat smaller. Organization is your choice since my private projects are structured completely differently anyways, though `gltfwrapper.cpp` containing all the code seems sensible based on your comments.

Reply:
I think having the loading & rendering code in the same file is fine.

I guess to me the cleanest (in the sense of fitting current Lobster) structure would be:

gltf_scene.h  // All the structs below that contain the scene loaded from a FlatBuffer. Depends as much as possible on glinterface.h. Has prototype decls for loading & rendering functions.

gltf.cpp  // Implements everything.

graphics.cpp  // (pre-existing) bindings for Lobster to load and render.

Reply:
Okay, no idea what you mean by half of that; let me see if I understand.

* Rename `flat_gltf.h` to `gltf_scene.h` and `gltfwrapper.cpp` to `gltf.cpp`
* Offload some of the meat from `gltf_scene.h` to `gltf.cpp`
* Replace my `LoadTexture` function with the Lobster equivalent

Reply:
Yes, and move the actual bindings to graphics.cpp.
And make other code use lobster equivalents as well. Like the code generating vertex buffers etc.

Reply:
If I use Lobster's VAO loader, won't I be stuck with Lobster's current shaders? I mean, `phong_anim` is not the ideal vertex format for glTF/PBR, and I'm not sure if Lobster's VAO loader is generalized enough yet to handle new vertex formats.

Reply:
Lobster's VAO code can deal with a variety of attributes, it is not fixed to one vertex format. They're also not tied to any particular shader, you can totally make your own.

Besides that, I see gltf and pbr as completely orthogonal issues, I don't see what they have to do with eachother. If rendering features are missing to support pbr, I am pretty sure we can add them.. either in this PR or maybe better later.

Reply:
Yeah, glTF and PBR meet at a perpendicular angle, much like vertex tangents and unlike vertex colors. But it's not like a model made with a PBR pipeline can't be kludged into Lobster's Phong pipeline, which is what I'm doing now. Once the damn thing renders (!) I will work on expanding Lobster's pipeline capabilities. Of course, I have avoided touching `graphics.cpp` up to this point, but you gave the ok so okay.